### PR TITLE
Add even more Ginkgo unit tests

### PR DESCRIPTION
**Pull request by qbarrand**
This PR is a port of #139 with Ginkgo.

/cc @pacevedom @ybettan @yevgeny-shnaidman
/ok-to-test

### CODE REVIEW

**Comment by qbarrand** [comment]
@ybettan @yevgeny-shnaidman PTAL

File: pkg/lifecycle/lifecycle_test.go [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+)
+
+var labels = map[string]string{"key": "value"}
+
+func TestLifecycle(t *testing.T) {
+	RegisterFailHandler(Fail)
+	RunSpecs(t, "Lifecycle Suite")
+}
+
+var _ = Describe("Lifecycle", func() {
```
Comment (yevgeny-shnaidman) [inline pkg/lifecycle/lifecycle_test.go]: Why do we need 2 levels of Describe here?  You can have one Describe with 2 Its per tested function, or 2 Describe per each function with one It under each

File: pkg/proxy/proxy_test.go [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+func TestProxy(t *testing.T) {
+	RegisterFailHandler(Fail)
+	RunSpecs(t, "Proxy Suite")
+}
+
+var _ = Describe("Proxy", func() {
```
Comment (yevgeny-shnaidman) [inline pkg/proxy/proxy_test.go]: same comment here - we don't need 2 levels of Describe

File: pkg/resource/resource_test.go [hunk 3]
```diff
@@ -0,0 +1,389 @@
...
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+func TestResource(t *testing.T) {
+	RegisterFailHandler(Fail)
+	RunSpecs(t, "Resource Suite")
+}
+
+var _ = Describe("Resource", func() {
+	Describe("IsNamespaced", func() {
```
Comment (yevgeny-shnaidman) [inline pkg/resource/resource_test.go]: 2 levels of Describes are not a very good practice. Lets' replace it with Context

File: pkg/scheme/scheme_test.go [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+func TestScheme(t *testing.T) {
+	RegisterFailHandler(Fail)
+	RunSpecs(t, "Scheme Suite")
+}
+
+var _ = Describe("Scheme", func() {
+	Describe("AddToScheme", func() {
```
Comment (yevgeny-shnaidman) [inline pkg/scheme/scheme_test.go]: Context instead of 2nd level describe

File: pkg/slice/slice_test.go [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+	"helm.sh/helm/v3/pkg/chart"
+)
+
+func TestSlice(t *testing.T) {
+	RegisterFailHandler(Fail)
+	RunSpecs(t, "Slice Suite")
+}
+
+var _ = Describe("Slice", func() {
+	Describe("Find", func() {
```
Comment (yevgeny-shnaidman) [inline pkg/slice/slice_test.go]: Context instead of second level Describe

File: pkg/state/state_test.go [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+	"github.com/openshift-psap/special-resource-operator/pkg/state"
+)
+
+func TestState(t *testing.T) {
+	RegisterFailHandler(Fail)
+	RunSpecs(t, "State Suite")
+}
+
+var _ = Describe("State", func() {
+	Describe("GenerateName", func() {
```
Comment (yevgeny-shnaidman) [inline pkg/state/state_test.go]: you don't need second Describe, It can be directly under first Describe

File: pkg/storage/storage_test.go [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+)
+
+var nsn = types.NamespacedName{Namespace: namespaceName, Name: resourceName}
+
+func TestStorage(t *testing.T) {
+	RegisterFailHandler(Fail)
+	RunSpecs(t, "Storage Suite")
+}
+
+var _ = Describe("Storage", func() {
```
Comment (yevgeny-shnaidman) [inline pkg/storage/storage_test.go]: What do you think about Describe per tested function? This file tests only storage.go, so we don't actually need to state it in a specific Describe

File: pkg/lifecycle/lifecycle_test.go (hunk 1 above)
Comment (qbarrand) [reply pkg/lifecycle/lifecycle_test.go]: Here, this allows me to define `BeforeEach` and `AfterEach` once, and it applies to both `GetPodFromDaemonSet` and `UpdateDaemonSetPods` tests. How can I do this if I remove the wrapping ` var _ = Describe("Lifecycle", func() {`?

File: pkg/proxy/proxy_test.go (hunk 2 above)
Comment (qbarrand) [reply pkg/proxy/proxy_test.go]: I agree here. These tests predate our agreement of one top-level `Describe` per tested function. Refactored.

File: pkg/resource/resource_test.go (hunk 3 above)
Comment (qbarrand) [reply pkg/resource/resource_test.go]: Removed ` var _ = Describe("Resource", func() {` as it's unnecessary here.

File: pkg/scheme/scheme_test.go (hunk 4 above)
Comment (qbarrand) [reply pkg/scheme/scheme_test.go]: Removed the top-level `Describe` here.

File: pkg/slice/slice_test.go (hunk 5 above)
Comment (qbarrand) [reply pkg/slice/slice_test.go]: Removed the top-level `Describe`.

File: pkg/state/state_test.go (hunk 6 above)
Comment (qbarrand) [reply pkg/state/state_test.go]: Removed the top-level `Describe`.

File: pkg/storage/storage_test.go (hunk 7 above)
Comment (qbarrand) [reply pkg/storage/storage_test.go]: This allows me to have a common `BeforeEach` for multiple `Context`. How could I do it without the top-level `Describe`?

File: pkg/scheme/scheme_test.go (hunk 4 above)
Comment (yevgeny-shnaidman) [inline pkg/scheme/scheme_test.go]: There is still a second level Describe here

File: pkg/resource/resource.go [hunk 8]
```diff
@@ -105,7 +105,7 @@ func UpdateResourceVersion(req *unstructured.Unstructured, found *unstructured.U
 	if kind == "Service" {
 		clusterIP, fnd, err := unstructured.NestedString(found.Object, "spec", "clusterIP")
 		if err != nil || !fnd {
-			return fmt.Errorf("error or not found: %w", err)
+			return fmt.Errorf("error or clusterIP found: %w", err)
```
Comment (yevgeny-shnaidman) [inline pkg/resource/resource.go]: should it be:
"error or clusterIP not found"?

File: pkg/lifecycle/lifecycle_test.go [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+	optLabels = client.MatchingLabels(labels)
+
+	unstructuredMatcher     = gomock.AssignableToTypeOf(&unstructured.Unstructured{})
+	unstructuredListMatcher = gomock.AssignableToTypeOf(&unstructured.UnstructuredList{})
+)
+
+func TestLifecycle(t *testing.T) {
+	RegisterFailHandler(Fail)
+
+	BeforeEach(func() {
```
Comment (ybettan) [inline pkg/lifecycle/lifecycle_test.go:40]: When is this one actually running? Never saw a `BeforeEach` or `AfterEach` outside a `Describe`

Comment (qbarrand) [reply pkg/lifecycle/lifecycle_test.go:40]: Just like in a `Describe`, it's running before each of the suite's `It`. @yevgeny-shnaidman came up with this suggestion to have `BeforeEach` when all tests are top-level `Describe`s.

File: pkg/scheme/scheme_test.go (hunk 4 above)
Comment (qbarrand) [reply pkg/scheme/scheme_test.go]: Not sure I missed this one. Fixed - sorry.

File: pkg/resource/resource.go (hunk 8 above)
Comment (qbarrand) [reply pkg/resource/resource.go]: Good catch, thanks - fixed.

**Pull request merged**